mlflow/utils/autologging_utils.py

```diff
@@ -275,3 +282,202 @@ def batch_metrics_logger(run_id):
     batch_metrics_logger = BatchMetricsLogger(run_id)
     yield batch_metrics_logger
     batch_metrics_logger._purge()
+
+
+def autologging_integration(name):
+    """
+    Wraps an autologging function in order to store its configuration arguments. This enables
+    patch functions to broadly obey certain configurations (e.g., disable=True) without
+    requiring specific logic to be present in each autologging integration. **All autologging
+    integrations should be decorated with this wrapper.**
+    """
+
+    AUTOLOGGING_INTEGRATIONS[name] = {}
+
+    def wrapper(_autolog):
+        def autolog(*args, **kwargs):
+            AUTOLOGGING_INTEGRATIONS[name] = kwargs
+            _autolog(**kwargs)
+
+        wrapped_autolog = functools.wraps(_autolog)(autolog)
+        return wrapped_autolog
+
+    return wrapper
+
+
+def safe_patch(autologging_integration, destination, function_name, function):
+    """
+    Patches the specified `function_name` on the specified `destination` class for autologging
+    purposes, replacing its implementation with an error-safe copy of the specified `function`.
+
+    :param autologging_integration: The name of the autologging integration associated with the
+                                    patch.
+    :param destination: The Python class on which the patch function is being defined.
+    :param function_name: The name of the function to patch on the specified `destination` class.
+    :param function: The patch function to apply. The first argument to this function should be
+                     reserved for an `original` method argument representing the underlying /
+                     original function. Subsequent arguments should be identical to those of the
+                     original function being patched.
+    """
+
+    def patched_train(*args, **kwargs):
```

Comment:
`patched_train` is a misnomer here, since not all patched methods corresponding to training routines